src/view/tree_structure_tracking.rs

```diff
@@ -0,0 +1,91 @@
+use crate::{
+    id::Id,
+    view::ElementsSplice,
+    widget::{ChangeFlags, Pod},
+};
+use xilem_core::VecSplice;
+
+use super::Cx;
+
+#[derive(Clone, Copy, Debug)]
+pub enum SpliceMutation {
+    Add(Id),
+    Change(Id),
+    Delete(usize),
+    Skip(usize),
+}
+
+/// An ElementsSplice that monitors the tree structure by maintaining a mutation log
+pub struct TreeStructureTrackingSplice<'a, 'b, 'c> {
+    current_child_id: Option<Id>,
+    splice: VecSplice<'a, 'b, Pod>,
+    mutations: &'c mut Vec<SpliceMutation>,
+}
+
+impl<'a, 'b, 'c> TreeStructureTrackingSplice<'a, 'b, 'c> {
+    pub fn new(
+        elements: &'a mut Vec<Pod>,
+        scratch: &'b mut Vec<Pod>,
+        mutations: &'c mut Vec<SpliceMutation>,
+    ) -> Self {
+        mutations.clear();
+        Self {
+            splice: VecSplice::new(elements, scratch),
+            current_child_id: None,
+            mutations,
+        }
+    }
+}
+
+impl<'a, 'b, 'c> ElementsSplice for TreeStructureTrackingSplice<'a, 'b, 'c> {
+    fn push(&mut self, element: Pod, cx: &mut Cx) {
+        self.mutations.push(SpliceMutation::Add(element.id()));
+        cx.apply_children_tree_structure_mutations(element.id());
+        self.splice.push(element);
+    }
```

Comment:
```suggestion
    fn push(&mut self, element: Pod, cx: &mut Cx) {
        if let Some(parent_id) = cx.id_path().last() {
          cx.tree_structure.append_child(parent_id, element.id);
        }
        self.splice.push(element);
    }
```

See my longer comment.